### PR TITLE
feat(instance): resource user data

### PR DESCRIPTION
**Pull request by Monitob**

### CODE REVIEW

File: docs/resources/instance_user_data.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+---
+page_title: "Scaleway: scaleway_instance_user_data"
+description: |-
+Manages Scaleway Compute Instance User Data.
+---
+
+# scaleway_instance_server
```
Comment (remyleone) [inline docs/resources/instance_user_data.md:7]:
```suggestion
# scaleway_instance_user_data
```

File: docs/resources/instance_user_data.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+Creates and manages Scaleway Compute Instance User Data values.
+
+User data is a key value store API you can use to provide data from and to your server without authentication. It is the mechanism by which a user can pass information contained in a local file to an Instance at launch time.
+
+The typical use case is to pass something like a shell script or a configuration file as user data.
```
Comment (remyleone) [inline docs/resources/instance_user_data.md:13]: such as cloud-init for instance

File: scaleway/resource_instance_user_data.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+		},
+		SchemaVersion: 0,
+		Schema: map[string]*schema.Schema{
+			"server_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "The ID of the server",
+				ValidateFunc: validationUUIDWithLocality(),
+			},
+			"key": {
```
Comment (remyleone) [inline scaleway/resource_instance_user_data.go:38]: Why not use an HCL maps: https://www.terraform.io/language/expressions/types#maps-objects ?

File: scaleway/resource_instance_user_data_test.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+EOF
+				}
+
+				resource "scaleway_instance_server" "main" {
+					image = "ubuntu_focal"
+					type  = "DEV1-S"
+				}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("scaleway_instance_user_data.main", "key", "cloud-init"),
```
Comment (remyleone) [inline scaleway/resource_instance_user_data_test.go:35]: Could you check that the value present on the server is correct? Could you also check that the value stored in the state is correct?

File: go.mod [hunk 5]
```diff
@@ -5,13 +5,13 @@ require (
...
 	github.com/dustin/go-humanize v1.0.0
-	github.com/google/go-cmp v0.5.8
+	github.com/google/go-cmp v0.5.9
 	github.com/hashicorp/aws-sdk-go-base v1.1.0
 	github.com/hashicorp/awspolicyequivalence v1.6.0
 	github.com/hashicorp/go-cty v1.4.1-0.20200414143053-d3edf31b6320
 	github.com/hashicorp/go-retryablehttp v0.7.1
 	github.com/hashicorp/terraform-plugin-log v0.7.0
-	github.com/hashicorp/terraform-plugin-sdk/v2 v2.21.0
+	github.com/hashicorp/terraform-plugin-sdk/v2 v2.23.0
```
[the remark this answers is not in the document]
Comment (Monitob) [reply go.mod]: commit dropped. I'll do it with a dedicated PR

File: go.mod [hunk 6]
```diff
@@ -5,13 +5,13 @@ require (
 	github.com/dnaeon/go-vcr v1.2.0
 	github.com/docker/docker v20.10.17+incompatible
 	github.com/dustin/go-humanize v1.0.0
-	github.com/google/go-cmp v0.5.8
+	github.com/google/go-cmp v0.5.9
```
[the remark this answers is not in the document]
Comment (Monitob) [reply go.mod]: commit dropped. I'll do it with a dedicated PR

File: docs/resources/instance_user_data.md (hunk 2 above)
Comment (Monitob) [reply docs/resources/instance_user_data.md:13]: Yes, I've taken some documentation already present on the resource instance server

File: scaleway/resource_instance_user_data.go (hunk 3 above)
Comment (Monitob) [reply scaleway/resource_instance_user_data.go:38]: Taking this approach we allow the client to update key by key instead update all the user-data.

File: scaleway/resource_instance_user_data_test.go (hunk 4 above)
Comment (Monitob) [reply scaleway/resource_instance_user_data_test.go:39]: I've just added the required test.

File: docs/resources/instance_user_data.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+- `zone` - (Defaults to [provider](../index.md#zone) `zone`) The [zone](../guides/regions_and_zones.md#zones) in which the server should be created.
+
+~> **Important:**   Use the `cloud-init` key to use [cloud-init](https://cloudinit.readthedocs.io/en/latest/) on your instance.
+  You can define values using:
+    - string
+    - UTF-8 encoded file content using [file](https://www.terraform.io/language/functions/file)
+
+## Import
+
+User data can be imported using the `{zone}/{key}/{id}`, e.g.
```
Comment (Codelax) [inline docs/resources/instance_user_data.md:73]: specify `server_id`

File: scaleway/resource_instance_user_data.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	server, err := waitForInstanceServer(ctx, instanceAPI, zone, id, d.Timeout(schema.TimeoutRead))
+	if err != nil {
+		if is404Error(err) {
+			d.SetId("")
+			return nil
+		}
+		return nil
+	}
```
Comment (Codelax) [inline scaleway/resource_instance_user_data.go:102]: Shouldn't we fail if instance does not exist ?

File: scaleway/resource_instance_user_data.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+	if v, ok := d.GetOk("zone"); ok {
+		deleteUserData.Zone = scw.Zone(v.(string))
+	}
+
+	err = instanceAPI.DeleteServerUserData(deleteUserData, scw.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
```
Comment (Codelax) [inline scaleway/resource_instance_user_data.go:195]: Can we check for a specific error if user data does not exist ?

**Review by Codelax — CHANGES_REQUESTED** [review]

**Pull request merged**